### PR TITLE
Fix Promise interfaces

### LINKED ISSUE #11094: Promise can be resolved with undefined but not reflected in type
TS 2.0.3

``` ts
    new Promise<string>(resolve => {
        true ? resolve('foo') :  resolve()
    })
```

This returns a `Promise<string>`, however it is possible to call `resolve` with no value, meaning the return type is actually `Promise<string | undefined>` at runtime.

I'm not sure whether the return type should in fact be inferred as `Promise<string | undefined>` or whether the empty `resolve()` call should throw a compile error.

### PR DESCRIPTION
**Pull request by falsandtru**
Fixes #11094, #3356, #21867.

@mhegazy @rbuckton Can you include this to the next release?

### CODE REVIEW

**Comment by falsandtru** [comment]
@rbuckton @mhegazy @DanielRosenwasser @ahejlsberg Can you merge this? This fix has no bad effect.

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1273,7 +1273,9 @@ declare type PropertyDecorator = (target: Object, propertyKey: string | symbol)
 declare type MethodDecorator = <T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>) => TypedPropertyDescriptor<T> | void;
 declare type ParameterDecorator = (target: Object, propertyKey: string | symbol, parameterIndex: number) => void;
 
-declare type PromiseConstructorLike = new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void) => PromiseLike<T>;
+interface PromiseConstructorLike {
```
Comment (rbuckton) [inline src/lib/es5.d.ts]: I don't think this specific change is necessary. `PromiseConstructorLike` is only used by the type system to determine if an explicit `Promise`-like return type in a down-level ES5 async function can be used as a `Promise` constructor. Changing the type could adversely affect people using this behavior.

File: src/lib/es2015.promise.d.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ interface PromiseConstructor {
      * a resolve callback used resolve the promise with a value or the result of another promise,
      * and a reject callback used to reject the promise with a provided reason or error.
      */
-    new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
+    new <T>(executor: (resolve: [T] extends [void] ? (value?: T | PromiseLike<T>) => void : (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
```
Comment (rbuckton) [inline src/lib/es2015.promise.d.ts]: Is the goal to only allow `value` to be optional if `T` is exactly `void`? If `T` is `void | number` you end up with the second definition (with a non-optional `value`) for `(value: void | number | PromiseLike<void | number>) => void`.

You could try this instead:
```ts
resolve: [void] extends [T] ? (value?: T | PromiseLike<T>) => void : (value: T | PromiseLike<T>) => void
```

By inverting the condition, you would get cases where `void` is assignable (such as when `T` is a union including `void`).

Comment (falsandtru) [reply src/lib/es2015.promise.d.ts]: It seems a nice idea. I'll do it.

File: src/lib/es5.d.ts (hunk 1 above)
Comment (falsandtru) [reply src/lib/es5.d.ts]: Done.

File: src/lib/es2015.promise.d.ts (hunk 2 above)
Comment (falsandtru) [reply src/lib/es2015.promise.d.ts]: Done.

Comment (falsandtru) [reply src/lib/es2015.promise.d.ts]: I realized `[void] extends [T]` doesn't work with `undefined`. So I reverted it.

Comment (rbuckton) [reply src/lib/es2015.promise.d.ts]: That's unfortunate, but its better than nothing I suppose.

Comment (rbuckton) [reply src/lib/es2015.promise.d.ts]: Have you tried this without the conditional type? This should work as well and is probably easier to read:

```ts
    new <T extends void>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
    new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;

```

Comment (falsandtru) [reply src/lib/es2015.promise.d.ts]: As I mentioned at https://github.com/Microsoft/TypeScript/pull/21867#issuecomment-375087677, it is not extensible.

**Comment by falsandtru** [comment]
@rbuckton Ah, I realized a problem now. In the following case, this change make an unfavorable effect.

```ts
class C<T> extends Promise<T> {
  constructor() {
    super(resolve => resolve(0 as any)); // error, can't call it.
  }
}
```

**Comment by weswigham** [comment]
@rbuckton Our RWC tests show this as kinda breaky, for a few reasons.
1. Generic promises may no longer yield callable `resolve` or `reject`, as the conditional type is not resolved,  and you get something like `((value?: A | PromiseLike<A> | undefined) => void) | ((value: A | PromiseLike<A>) => void)` as the apparent type, which do not unify and is not callable under current rules.
2. Promises where inference has failed (eg, a simple `new Promise` without type annotation), have a type of `{}`, which, by virtue of not being void, _require_ an argument to `resolve`, which very much does not jibe with real-world use - in fact, in many cases `new Promise()` is used like a `Promise<void>` (and why wouldn't you).

**Comment by rbuckton** [comment]
I'm reverting this change for now.

**Pull request merged**